File: CurrencyBank.WPF/Services/AccountService.cs
﻿using CurrencyBank.WPF.Dto;
using CurrencyBank.WPF.Models;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace CurrencyBank.WPF.Services
{
    public class AccountService
    {
        private static readonly HttpClient _client = new HttpClient();
        private const string _baseUrl = "http://localhost:5000/api/account";

        public async Task<HttpResponseMessage> GetAccountsForUser(string token)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            HttpResponseMessage response = await _client.GetAsync(_baseUrl + "/getAccounts");

            return response;
        }

        public async Task<HttpResponseMessage> Withdrawal(string token, int accountId, decimal ammount)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            HttpResponseMessage response = await _client.PutAsync(_baseUrl + "/cashout?accountId=" + accountId + "&ammount="+ ammount, null);

            return response;
        }

        public async Task<HttpResponseMessage> OpenAccount(string token, AccountToCreateDto accountToCreateDto)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            var data = new
            {
                Currency = accountToCreateDto.Currency
            };

            HttpResponseMessage response = await _client.PostAsJsonAsync(_baseUrl + "/create", data);

            return response;
        }

        public async Task<HttpResponseMessage> DeleteAccount(string token, int id)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            HttpResponseMessage response = await _client.DeleteAsync(_baseUrl + "/deleteAccount?accountId=" + id);

            return response;
        }

        public async Task<HttpResponseMessage> ExchangeMoney(string token, int sourceAccountId, int destinationAccountId, decimal ammount)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            string url = $"{_baseUrl}/exchange?sourceAccountId={sourceAccountId}&destinationAccountId={destinationAccountId}&ammount={ammount}";

            HttpResponseMessage response = await _client.PutAsync(url, null);

            return response;
        }

        public async Task<HttpResponseMessage> TransferMoney(string token, int principalAccountId, string destinationAccountNumber, decimal ammount)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            string url = $"{_baseUrl}/transferMoney?principalAccountId={principalAccountId}&destinationAccountNumber={destinationAccountNumber}&ammount={ammount}";

            HttpResponseMessage response = await _client.PostAsync(url, null);

            return response;
        }

        public async Task<HttpResponseMessage> CashInMoney(string token, int accountId, decimal ammount)
        {
            _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
            _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            string url = $"{_baseUrl}/cashIn?accountId={accountId}&ammount={ammount}";

            HttpResponseMessage response = await _client.PutAsync(url, null);

            return response;
        }
    }
}

File: currency-bank-web-app/src/_services/account.service.ts
import { IAccount } from "./../interfaces/IAccount";
import { INewAccount } from "./../interfaces/NewAccount/INewAccount";
import { _post, _get, _delete, _put } from "../utils/ApiRequest";
import { IHistoryAccounts } from "../interfaces/IHistoryAccounts";

export const accountService = {
  getAllAccountsForCurrentUser,
  getHistoryForUserAccounts,
  openNew,
  deleteAccount,
  cashOut,
  cashIn,
  transfer,
  exchange,
};

function getAllAccountsForCurrentUser() {
  return _get<IAccount[]>(`account/getAccounts`);
}

function getHistoryForUserAccounts() {
  return _get<IHistoryAccounts[]>(`account/getHistory`);
}

function openNew(newAccount: INewAccount) {
  return _post<IAccount>(`account/create`, newAccount);
}

function deleteAccount(accountId: number) {
  return _delete<string>(`account/deleteAccount?accountId=${accountId}`);
}

function cashOut(accountId: number, ammount: number) {
  return _put<IAccount>(
    `account/cashout?accountId=${accountId}&ammount=${ammount}`
  );
}

function cashIn(accountId: number, ammount: number) {
  return _put<IAccount>(
    `account/cashIn?accountId=${accountId}&ammount=${ammount}`
  );
}

function transfer(
  accountId: number,
  destAccountNumber: string,
  ammount: number
) {
  return _post<IAccount[]>(
    `account/transferMoney?principalAccountId=${accountId}&destinationAccountNumber=${destAccountNumber}&ammount=${ammount}`
  );
}

function exchange(
  srcAccountId: number,
  destAccountId: number,
  ammount: number
) {
  return _put<IAccount[]>(
    `account/exchange?sourceAccountId=${srcAccountId}&destinationAccountId=${destAccountId}&ammount=${ammount}`
  );
}
